File: Sorting/Selection Sort/SelectionSort.js
// time complexity => o(n^2)

let arr = [-1,-10, 50,20,2,3,54,32,-3];

function selectionSort(arr){
    for(let i = 0; i < arr.length - 1; i++){ // just go to last second element
        let min_el_idx = findMinEl(arr, i);

        if(min_el_idx !== i){
            // min elment and first el of unsorted region are diff then swap 
            let temp = arr[i];
            arr[i] = arr[min_el_idx];
            arr[min_el_idx] = temp;
        }
    }

}

function findMinEl(arr, i){
    let min_el_idx = i;
    
    for(let j = i + 1; j < arr.length; j++){
        if(arr[min_el_idx] > arr[j]){
            min_el_idx = j;
        }
    }

    return min_el_idx;

}

selectionSort(arr);
console.log(arr);

File: Extra Practice/Self Practice/Sorting/selectionSort.java


public class selectionSort {

    public static void selectionsort(int[] arr){
        int n = arr.length;

        for(int i = 0; i < n-1; i++){
            int min_idx = findMinElement(arr, i);
            if(min_idx != i){
                int temp = arr[i];
                arr[i] = arr[min_idx];
                arr[min_idx] =  temp;
            }
        }
    }

    public static int findMinElement(int[] arr, int i){
        int min_el_idx = i;
        for(int j = i+1 ; j < arr.length; j++){
            if(arr[j] < arr[min_el_idx]){
                min_el_idx = j;
            }
        }
        return min_el_idx;
    }


    public static void main(String[] args) {
        int[] arr = {1,3,4,5,6,10,2,3,1};
        selectionsort(arr);
        for(int element: arr){
            System.out.print(element + " ");
        }
    }
}


